odl/trafos/convolution.py

```diff
@@ -0,0 +1,804 @@
+            ``kernel`` is callable.
+
+        Returns
+        -------
+        kernel : `DiscreteLpVector`
+            Element in an adequate space.
+        range : `DiscreteLp`
+            Range of the convolution operator.
+        """
+        if isinstance(kernel, domain.element_type):
+            if kernel.ndim != domain.ndim:
+                raise ValueError('`kernel` has dimension {}, expected {}'
+                                 ''.format(kernel.ndim, domain.ndim))
+            # Set the range equal to the domain shifted by the midpoint of
+            # the kernel space.
+            kernel_shift = kernel.space.partition.midpoint
+            range = uniform_discr_fromdiscr(
+                domain, min_corner=domain.min_corner + kernel_shift)
+            return kernel, range
+
+        else:
+            if callable(kernel):
+                # Use "natural" kernel space, which is the zero-centered
+                # version of the domain.
+                extent = domain.partition.extent()
+                std_kernel_space = uniform_discr_fromdiscr(
+                    domain, min_corner=-extent / 2)
+                kernel = std_kernel_space.element(kernel, **kernel_kwargs)
+                range = domain
+                return kernel, range
+            else:
+                # Make a zero-centered space with the same cell sides as
+                # domain, but shape according to the kernel.
+                kernel = np.asarray(kernel, dtype=domain.dtype)
+                nsamples = kernel.shape
+                extent = domain.cell_sides * nsamples
+                kernel_space = uniform_discr_fromdiscr(
+                    domain, min_corner=-extent / 2, max_corner=extent / 2,
+                    nsamples=nsamples)
+                kernel = kernel_space.element(kernel)
+                range = domain
+                return kernel, range
+
+    @staticmethod
+    def _get_resamp_csides(resample, ker_csides, dom_csides):
+        """Return cell sides for kernel and domain resampling."""
+        new_ker_csides = []
+        new_dom_csides = []
+        for resamp, ks, ds in zip(resample, ker_csides, dom_csides):
+
+            if np.isclose(ks, ds):
+                # Keep old if both csides are the same
+                new_ker_csides.append(ks)
+                new_dom_csides.append(ds)
+            else:
+                # Pick the coarser one if 'down' or the finer one if 'up'
+                if ((ks < ds and resamp == 'up') or
+                        (ks > ds and resamp == 'down')):
+                    new_ker_csides.append(ks)
+                    new_dom_csides.append(ks)
+                else:
+                    new_ker_csides.append(ds)
+                    new_dom_csides.append(ds)
+
+        return new_ker_csides, new_dom_csides
+
+    @property
+    def impl(self):
+        """Implementation of this operator."""
+        return self._impl
+
+    @property
+    def resample(self):
```

Comment:
`resampling` perhaps? `resample` sounds like a function that will do something, i.e. this feels natural

```python
conv.resample(something)
```